File: programa.py
"""Juego para adivinar una palabra letra por letra.

El programa elige una palabra al azar. El usuario tiene hasta 5 errores para
descubrir todas las letras que forman la palabra.
"""

import random


def pedir_un_caracter(mensaje, letras_registradas):
    while True:
        entrada = input(mensaje).lower()

        if len(entrada) != 1:
            print("Error: Por favor, ingresa solo un caracter.")
        elif entrada in letras_registradas:
            print("Error: Por favor, ingresa una letra no repetida.")
        else:
            return entrada


PALABRAS = [
    "uno",
    "dos",
    "tres",
    "cuatro",
    "cinco",
    "seis",
    "siete",
    "ocho",
    "nueve",
    "diez",
]

CLAVE = random.choice(PALABRAS)
letras_registradas = []
letras_correctas = []
intentos = 5
gana = False

while intentos > 0 and not gana:
    print(f"\n\t ..::  INTENTOS RESTANTES: {intentos}  ::..")

    letra = pedir_un_caracter("Introduce un solo caracter: ", letras_registradas)
    letras_registradas.append(letra)

    if letra in CLAVE:
        letras_correctas.append(letra)
        print(f" LA LETRA {letra} SI ESTA EN LA PALABRA CLAVE ✅ ")

        gana = True

        for letra_clave in CLAVE:
            if letra_clave not in letras_registradas:
                gana = False
                break
    else:
        intentos -= 1
        print(f" LA LETRA {letra} NO ESTA EN LA PALABRA CLAVE ❌ ")

    palabra_mostrada = ""
    for letra_clave in CLAVE:
        if letra_clave in letras_correctas:
            palabra_mostrada += letra_clave
        else:
            palabra_mostrada += "_"

    print(f"Palabra: {palabra_mostrada}")

print("=" * 50)

if gana:
    print(f"\n\t FELICIDADES: LA PALABRA ES -> {CLAVE}")
else:
    print(f"\n\t LA PALABRA ES -> {CLAVE}")

print("\n\t ..::   PROGRAMA FINALIZADO   ::..")

File: Programa.java
import java.util.ArrayList;
import java.util.List;
import java.util.Random;
import java.util.Scanner;

public class Programa {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        String[] palabras = {
            "uno",
            "dos",
            "tres",
            "cuatro",
            "cinco",
            "seis",
            "siete",
            "ocho",
            "nueve",
            "diez"
        };

        Random random = new Random();
        String clave = palabras[random.nextInt(palabras.length)];

        List<Character> letrasRegistradas = new ArrayList<>();
        List<Character> letrasCorrectas = new ArrayList<>();

        int intentos = 5;
        boolean gana = false;

        while (intentos > 0 && !gana) {
            System.out.println("\n\t ..::  INTENTOS RESTANTES: " + intentos + "  ::..");

            char letra = pedirUnCaracter(
                scanner,
                "Introduce un solo caracter: ",
                letrasRegistradas
            );

            letrasRegistradas.add(letra);

            if (clave.indexOf(letra) != -1) {
                letrasCorrectas.add(letra);
                System.out.println(" LA LETRA " + letra + " SI ESTA EN LA PALABRA CLAVE ✅ ");

                gana = true;

                for (int i = 0; i < clave.length(); i++) {
                    char letraClave = clave.charAt(i);

                    if (!letrasCorrectas.contains(letraClave)) {
                        gana = false;
                        break;
                    }
                }
            } else {
                intentos--;
                System.out.println(" LA LETRA " + letra + " NO ESTA EN LA PALABRA CLAVE ❌ ");
            }

            String palabraMostrada = "";

            for (int i = 0; i < clave.length(); i++) {
                char letraClave = clave.charAt(i);

                if (letrasCorrectas.contains(letraClave)) {
                    palabraMostrada += letraClave;
                } else {
                    palabraMostrada += "_";
                }
            }

            System.out.println("Palabra: " + palabraMostrada);
        }

        System.out.println("=".repeat(50));

        if (gana) {
            System.out.println("\n\t FELICIDADES: LA PALABRA ES -> " + clave);
        } else {
            System.out.println("\n\t LA PALABRA ES -> " + clave);
        }

        System.out.println("\n\t ..::   PROGRAMA FINALIZADO   ::..");

        scanner.close();
    }

    public static char pedirUnCaracter(
        Scanner scanner,
        String mensaje,
        List<Character> letrasRegistradas
    ) {
        while (true) {
            System.out.print(mensaje);
            String entrada = scanner.nextLine().toLowerCase();

            if (entrada.length() != 1) {
                System.out.println("Error: Por favor, ingresa solo un caracter.");
            } else {
                char letra = entrada.charAt(0);

                if (letrasRegistradas.contains(letra)) {
                    System.out.println("Error: Por favor, ingresa una letra no repetida.");
                } else {
                    return letra;
                }
            }
        }
    }
}
